app/src/main/kotlin/com/mapzen/erasermap/view/MultiModalDirectionListAdapter.kt

```diff
@@ -0,0 +1,423 @@
+  }
+
+  private fun listItemForLastInstruction(instructionGroup: InstructionGroup): ListRowItem {
+    val index = instructionGroup.instructions.size-1
+    val item = ListRowItem(ListRowType.ARRIVAL, R.layout.icon_title_row)
+    val instruction = instructionGroup.instructions[index]
+    item.iconId = DisplayHelper.getRouteDrawable(context, instruction.turnInstruction)
+    item.title = instruction.getHumanTurnInstruction()
+    return item
+  }
+
+  private fun setIconTitleRow(position: Int, holder: IconTitleViewHolder) {
+    val item = listItems[position]
+    holder.titleView.text = item.title
+    holder.iconView.setImageResource(item.iconId as Int)
+  }
+
+  private fun setPedestrianRow(position: Int, holder: PedestrianViewHolder) {
+    val listItem = listItems[position]
+    val instructionGroup = listItem.extra as InstructionGroup
+    holder.totalDistanceView.distanceInMeters = instructionGroup.totalDistance
+    holder.totalTimeView.timeInMinutes = instructionGroup.totalTime / SEC_PER_MIN
+    if (listItem.expanded) {
+      holder.distanceTimeContainer.openArrow()
+      holder.instructionsContainer.visibility = View.VISIBLE
+    } else {
+      holder.distanceTimeContainer.closeArrow()
+      holder.instructionsContainer.visibility = View.GONE
+    }
+    holder.distanceTimeContainer.setOnClickListener {
+      listItem.expanded = !listItem.expanded
+      notifyDataSetChanged()
+    }
+
+    for (i in 0..holder.instructionsContainer.childCount - 1) {
+      recycler.queueView(holder.instructionsContainer.getChildAt(i), VIEW_TYPE_INSTRUCTION_ROW)
+    }
+    holder.instructionsContainer.removeAllViews()
+    if (listItem.expanded) {
+      for (i in 0..instructionGroup.instructions.size - 1) {
+        val instruction = instructionGroup.instructions[i]
+        if (INSTRUCTION_TYPES_NOT_SHOWN.contains(instruction.turnInstruction)) {
+          continue
+        }
+        val distance = instruction.distance
+        var instructionRow = recycler.dequeueView(VIEW_TYPE_INSTRUCTION_ROW)
+        if (instructionRow == null) {
+          instructionRow = View.inflate(context, R.layout.instruction_row, null)
+        }
+
+        var iconId = DisplayHelper.getRouteDrawable(context, instruction.turnInstruction)
+        if (instruction.getTravelMode() == TravelMode.TRANSIT) {
+          iconId = multiModalHelper.getTransitIcon(instructionGroup.travelType)
+        }
+
+        val iconView = instructionRow?.findViewById(R.id.icon) as ImageView
+        val titleView = instructionRow?.findViewById(R.id.title) as TextView
+        val distanceView = instructionRow?.findViewById(R.id.distance) as DistanceView
+
+        iconView.setImageResource(iconId)
+        titleView.text = instruction.getHumanTurnInstruction()
+        distanceView.distanceInMeters = distance
+
+        holder.instructionsContainer.addView(instructionRow)
+      }
+    }
+
+  }
+
+  private fun setTransitRow(position: Int, holder: TransitViewHolder) {
```

Comment:
This function seems to do a lot. How hard would it be to refactor it into some smaller utility functions? Maybe a couple getters for some of the model related things?